Public/Src/Cache/DistributedCache.Host/Service/Internal/DualServerClientContentStore.cs

```diff
@@ -0,0 +1,172 @@
+            Contract.Assert(!string.IsNullOrWhiteSpace(cacheName));
+            Contract.Assert(!string.IsNullOrWhiteSpace(ringId));
+            Contract.Assert(!string.IsNullOrWhiteSpace(stampId));
+
+            var redisConnectionString = Environment.GetEnvironmentVariable(EnvironmentConnectionStringProvider.RedisConnectionStringEnvironmentVariable);
+
+            var localCasSettings = LocalCasSettings.Default(maxQuotaMB, dataRootPath.Path, cacheName, (uint)grpcPort);
+            localCasSettings.ServiceSettings.ScenarioName = scenarioName;
+
+            if (dataRootPath.IsLocal)
+            {
+                localCasSettings.DrivePreferenceOrder.Add(AbsolutePath.RootPath(dataRootPath.DriveLetter).Path);
+            }
+
+            var contentServerFactory = new ContentServerFactory(new DistributedCacheServiceArguments(
+                logger,
+                copier,
+                pathTransformer,
+                new TestDistributedCacheServiceHost(),
+                new HostInfo(stampId, ringId, new List<string>()),
+                default(CancellationToken),
+                dataRootPath.Path,
+                new DistributedCacheServiceConfiguration(
+                    localCasSettings,
+                    DistributedContentSettings.CreateEnabled(new Dictionary<string, string>() { { stampId, redisConnectionString } }, true)),
+                string.Empty));
+            _localContentServer = contentServerFactory.Create();
+
+            _serviceClientContentStore = new ServiceClientContentStore(
+                logger,
+                fileSystem,
+                new ServiceClientContentStoreConfiguration(cacheName, new ServiceClientRpcConfiguration(grpcPort), scenarioName));
+        }
+
+        /// <inheritdoc />
+        public CreateSessionResult<IReadOnlyContentSession> CreateReadOnlySession(Context context, string name, ImplicitPin implicitPin)
+        {
+            return _serviceClientContentStore.CreateReadOnlySession(context, name, implicitPin);
+        }
+
+        /// <inheritdoc />
+        public CreateSessionResult<IContentSession> CreateSession(Context context, string name, ImplicitPin implicitPin)
+        {
+            return _serviceClientContentStore.CreateSession(context, name, implicitPin);
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            _serviceClientContentStore?.Dispose();
+            _localContentServer?.Dispose();
+        }
+
+        /// <inheritdoc />
+        public Task<GetStatsResult> GetStatsAsync(Context context)
+        {
+            return _serviceClientContentStore.GetStatsAsync(context);
+        }
+
+        /// <inheritdoc />
+        public async Task<BoolResult> ShutdownAsync(Context context)
+        {
+            await _serviceClientContentStore.ShutdownAsync(context).ThrowIfFailure();
+            return await _localContentServer.ShutdownAsync(context).ThrowIfFailure();

```

Comment:
>_localContentServer [](start = 25, length = 19)

I can't understand the nullability semantics here. I can see that some members are using null-safe access but some others not.

And it seems that these fields should not be null and all ?. accesses are redundant and confusing...